### PR TITLE
develop JsonPath API

### PR DESCRIPTION
**Pull request by julien-truffaut**
Add Prism for JsonNull
Add at method to JsonPath to enable insert / deletion
Rename JsonPath.at to index
Add JsonPath methods for each Json primitive e.g. boolean, string, array etc ...
Add each, filterObj, filterArray to JsonPath
Add JsonPathTraversal for JsonPath backed by a `Traversal`

### CODE REVIEW

File: optics/src/main/scala/io/circe/optics/JsonPath.scala [hunk 1]
```diff
@@ -1,19 +1,36 @@
...
 import io.circe.optics.JsonObjectOptics._
 import io.circe.optics.JsonOptics._
 import monocle.{ Optional, Prism }
-import monocle.function.Index.index
+import monocle.function.{At, Index}
 import monocle.std.list._
 import scala.language.dynamics
 
 final case class JsonPath(json: Optional[Json, Json]) extends Dynamic {
+  final def Null: Optional[Json, Unit] = json composePrism jsonNull
```
Comment (julien-truffaut) [inline optics/src/main/scala/io/circe/optics/JsonPath.scala]: I am not sure if `Null` fits with the API

**Comment by julien-truffaut** [comment]
I am not sure why the test is failing, it passes locally.

File: optics/src/main/scala/io/circe/optics/JsonPath.scala [hunk 2]
```diff
@@ -1,19 +1,45 @@
...
-    JsonPath(json.composePrism(jsonObject).composeOptional(index(field)))
+    JsonPath(json.composePrism(jsonObject).composeOptional(Index.index(field)))
+
+  final def index(i: Int): JsonPath =
+    JsonPath(json.composePrism(jsonArray).composeOptional(Index.index(i)))
+
+  final def each: JsonTraversalPath =
+    JsonTraversalPath(json composeTraversal jsonDescendants)
+
+  final def arrayFilter(p: Int => Boolean): JsonTraversalPath =
```
Comment (julien-truffaut) [inline optics/src/main/scala/io/circe/optics/JsonPath.scala]: I don't know if it is very useful to have a filter on array index maybe it makes more sense to have a single filter that work object key. WDYT?

File: optics/src/main/scala/io/circe/optics/JsonPath.scala [hunk 3]
```diff
@@ -30,3 +56,42 @@ final case class JsonPath(json: Optional[Json, Json]) extends Dynamic {
 final object JsonPath {
   final val root: JsonPath = JsonPath(Optional.id)
 }
+
+final case class JsonTraversalPath(json: Traversal[Json, Json]) extends Dynamic {
```
Comment (julien-truffaut) [inline optics/src/main/scala/io/circe/optics/JsonPath.scala]: I am not attached to the name `JsonTraversalPath`, any other ideas?

**Comment by travisbrown** [comment]
Sorry for the long delay, @julien-truffaut—thanks for updating the names. Merging now…

**Comment by julien-truffaut** [comment]
np thanks for the comments

**Pull request merged**